File: SimpleEngine/MatrixHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace SimpleEngine
{
    public class MatrixHelper
    {
        mat44 mat;

        public MatrixHelper(mat44 mat)
        {
            this.mat = mat;
        }

        public MatrixHelper mult(mat44 m)
        {
            mat = mat * m;
            return this;
        }

        public MatrixHelper Rotate(AxisAngle aa)
        {
            return mult(mat44.FromAxisAngle(aa));
        }
        public MatrixHelper Rotate(quat q)
        {
            return mult(q.Conjugate.mat33.mat44);
        }

        public MatrixHelper Translate(vec3 t)
        {
            return mult(mat44.TranslationFor(t));
        }

        public vec3 transform(vec3 v)
        {
            return mat * v;
        }

        public mat33 mat33
        {
            get
            {
                return mat.mat33;
            }
        }

        public mat44 mat44
        {
            get
            {
                return mat;
            }
        }
    }
}

File: src/engine/matrixhelper.cc
#include "engine/matrixhelper.h"

#include "engine/mat33.h"
#include "engine/quat.h"
#include "engine/vec3.h"

namespace simple
{
    MatrixHelper::MatrixHelper(const simple::mat44& m)
        : matrix(m)
    {
    }

    MatrixHelper& MatrixHelper::Multiply(const simple::mat44& m)
    {
        matrix = matrix * m;
        return *this;
    }

    MatrixHelper& MatrixHelper::Rotate(const AxisAngle& aa)
    {
        return Multiply(mat44::FromAxisAngle(aa));
    }

    MatrixHelper& MatrixHelper::Rotate(const quat& q)
    {
        return Multiply(q.Conjugate().GetMatrix33().AsMat44());
    }

    MatrixHelper& MatrixHelper::Translate(const vec3& t)
    {
        return Multiply(mat44::FromTranslation(t));
    }

    vec3 MatrixHelper::GetTransform(const vec3& v) const
    {
        return matrix * v;
    }

    mat33 MatrixHelper::AsMat33() const
    {
        return matrix.AsMat33();
    }

    mat44 MatrixHelper::AsMat44() const
    {
        return matrix;
    }
}
